### PR TITLE
Replace the "Preselect" bool with an integer.

### PR DESCRIPTION
**Pull request by rchande**
Add a public "MatchPriority" class that defines 2 possible values.
Add a public "MatchPriority" class that defines 2 possible values. 
This change doesn't implement any different matching, but allows us to add the APIs in a less-breaking way.

### CODE REVIEW

**Comment by rchande** [comment]
Please review @CyrusNajmabadi @dotnet/roslyn-ide

**Comment by rchande** [comment]
@jasonmalinowski Should I be listing those APIs in Shipping for U3?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs [hunk 1]
```diff
@@ -26,64 +26,65 @@ protected AbstractCSharpCompletionProviderTests(CSharpTestWorkspaceFixture works
...
         {
-            return VerifyAtPositionAsync(code, position, ItemPartiallyWritten(expectedItemOrNull), usePreviousCharAsTrigger, expectedItemOrNull, expectedDescriptionOrNull, sourceCodeKind, checkForAbsence, experimental, glyph);
+            return VerifyAtPositionAsync(code, position, ItemPartiallyWritten(expectedItemOrNull), usePreviousCharAsTrigger, expectedItemOrNull, expectedDescriptionOrNull, sourceCodeKind, checkForAbsence, experimental, glyph, matchPriority);
         }
 
-        private async Task VerifyAtEndOfFileAsync(string code, int position, string insertText, bool usePreviousCharAsTrigger, string expectedItemOrNull, string expectedDescriptionOrNull, SourceCodeKind sourceCodeKind, bool checkForAbsence, bool experimental, int? glyph)
+        private async Task VerifyAtEndOfFileAsync(string code, int position, string insertText, bool usePreviousCharAsTrigger, string expectedItemOrNull, string expectedDescriptionOrNull, SourceCodeKind sourceCodeKind, bool checkForAbsence, bool experimental, int? glyph, int? matchPriority)
+
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs]: extra newline.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs [hunk 2]
```diff
@@ -465,5 +465,31 @@ enum Colors
 ";
             await VerifyItemExistsAsync(markup, "Colors");
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task NotAfterDot()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs]: This test probably shouldn't be in this PR.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 3]
```diff
@@ -58,6 +58,14 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
                     return;
                 }
 
+                // Don't show up within member access
+                // This previously worked because the type inferrer didn't work
+                // in member access expressions.
+                if (token.IsKind(SyntaxKind.DotToken))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs]: Should not be part of this PR.

File: src/Features/Core/Portable/Completion/MatchPriority.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /// augment or override the existing text-based matching.
+    /// </summary>
+    public static class MatchPriority
+    {
+        /// <summary>
+        /// The matching algorithm should give this item no special treatment.
+        /// 
+        /// Ordinary <see cref="CompletionProvider"/>s typically specify this.
+        /// </summary>
+        public const int Default = 0;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/MatchPriority.cs]: make this static readonly.

File: src/Features/VisualBasic/Portable/Completion/CompletionProviders/EnumCompletionProvider.vb [hunk 5]
```diff
@@ -33,7 +33,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.Providers
                                                                    Return m.Kind = SymbolKind.Field AndAlso
                                                                       DirectCast(m, IFieldSymbol).IsConst AndAlso
                                                                       m.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation)
-                                                               End Function))
+                                                               End Function).Concat(enumType))
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/CompletionProviders/EnumCompletionProvider.vb]: not as part of this PR.

File: src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionService.vb [hunk 6]
```diff
@@ -86,8 +86,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion
 
         Protected Overrides Function GetBetterItem(item As CompletionItem, existingItem As CompletionItem) As CompletionItem
             ' If one Is a keyword, And the other Is some other item that inserts the same text as the keyword,
-            ' keep the keyword (VB only)
-            If IsKeywordItem(existingItem) Then
+            ' keep the keyword (VB only), unless the other item is preselected
+            If IsKeywordItem(existingItem) AndAlso existingItem.Rules.MatchPriority >= item.Rules.MatchPriority Then
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/Completion/VisualBasicCompletionService.vb]: not as part of this PR.

**Comment by CyrusNajmabadi** [comment]
Ravi, can you undo any changes unrelated to the change of shape to the API?

You can leave in the test harness changes.  But we shouldn't be changing any other logic here.  Thanks!

**Comment by CyrusNajmabadi** [comment]
Paging @MattGertz 
We'd like to take this change now so we can prevent making a breaking change later on.

We're publishing our Completion API for Update3.  This is a small tweak to that API we're publishing to support a new scenario that will come in Dev15.  If we make this small tweak now then we won't need to make a breaking change (or figure out some other way to merge both APIs) at that time.

File: src/EditorFeatures/Core/Extensibility/Completion/CompletionHelper.cs [hunk 7]
```diff
@@ -297,11 +302,11 @@ protected int CompareMatches(PatternMatch match1, PatternMatch match2, Completio
             }
 
             // preselected items are prefered
-            if (item1.Rules.Preselect && !item2.Rules.Preselect)
+            if (comparePriorityBeforeCapitalization && item1.Rules.MatchPriority > item2.Rules.MatchPriority)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/Completion/CompletionHelper.cs]: Should remove the comparePriorityBeforeCapitalization  code as well.

File: src/EditorFeatures/VisualBasic/Completion/VisualBasicCompletionHelper.vb [hunk 8]
```diff
@@ -87,7 +87,7 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.Completion
                 End If
             End If
 
-            Return MyBase.IsBetterFilterMatch(item1, item2, filterText, trigger, filterReason, recentItems)
+            Return IsBetterFilterMatchWorker(item1, item2, filterText, recentItems, comparePriorityBeforeCapitalization:=True)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasic/Completion/VisualBasicCompletionHelper.vb]: Should probably not change this.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 9]
```diff
@@ -58,7 +58,7 @@ internal static class SymbolCompletionItem
                 filterText: filterText ?? (displayText.Length > 0 && displayText[0] == '@' ? displayText : symbols[0].Name),
                 sortText: sortText ?? symbols[0].Name,
                 glyph: glyph ?? symbols[0].GetGlyph(),
-                preselect: preselect,
+                matchPriority: matchPriority.GetValueOrDefault(),
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: nit: you should use GetValueOrDefault in the other places as well.

**Comment by MattGertz** [comment]
I'm all for being proactive. :-)  Approved pending the usual stuff.

File: src/EditorFeatures/VisualBasic/Completion/VisualBasicCompletionHelper.vb (hunk 8 above)
Comment (rchande) [inline src/EditorFeatures/VisualBasic/Completion/VisualBasicCompletionHelper.vb]: Removed.

File: src/EditorFeatures/Core/Extensibility/Completion/CompletionHelper.cs (hunk 7 above)
Comment (rchande) [inline src/EditorFeatures/Core/Extensibility/Completion/CompletionHelper.cs]: Removed.

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs (hunk 9 above)
Comment (rchande) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs]: Done.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs (hunk 1 above)
Comment (rchande) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AbstractCSharpCompletionProviderTests.cs]: Removed.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 10]
```diff
@@ -45,20 +45,20 @@ public sealed class CompletionItemRules
         /// <summary>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: Comment is wrong.  but i dont' care that much.  If you do change this, indicate that null means "i don't care", and then things with a higher match priority are picked over those that are lower or null.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 11]
```diff
@@ -68,26 +68,26 @@ public sealed class CompletionItemRules
         /// <param name="commitCharacterRules">Rules about which keys typed caused the completion item to be committed.</param>
         /// <param name="enterKeyRule">Rule about whether the enter key is passed through to the editor after the selected item has been committed.</param>
         /// <param name="formatOnCommit">True if the modified text should be formatted automatically.</param>
-        /// <param name="preselect">True if the related completion item should be initially selected.</param>
+        /// <param name="matchPriority">True if the related completion item should be initially selected.</param>
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: Same withthis comment.

**Comment by CyrusNajmabadi** [comment]
LGTM.  But i do think this property should be an ```int?``` not an int.  What do you think?

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by rchande** [comment]
I think we're going to leave it an `int` as we discussed on Slack.

**Comment by CyrusNajmabadi** [comment]
Sounds good.  Also @paulvanbrenk This is a breaking cahnge we're making in Update3.

**Comment by CyrusNajmabadi** [comment]
Note: it's a binary breaking change, but not a source break for you.  i.e. your code:

```
context.AddItem(CompletionItem.Create(displayText: entry.Name, sortText: entry.SortText, span: span, tags: tags));
```

Will continue to work if you recompile.  Is that something that you'll be able to do for the Update3 branch?

**Comment by CyrusNajmabadi** [comment]
@rchande Can you hold off merging on this until @paulvanbrenk confirms that TS won't have a problem with this?

**Comment by CyrusNajmabadi** [comment]
But, in the meantime, if you want to open this same PR against Master, that would be good :)

**Comment by CyrusNajmabadi** [comment]
Note: you can do this directly from github.  You don't need to make a new branch.

File: src/Features/Core/Portable/Completion/CompletionItemRules.cs [hunk 12]
```diff
@@ -91,6 +91,38 @@ public sealed class CompletionItemRules
...
+        /// <param name="preselect">True if the related completion item should be initially selected.</param>
+        /// <returns></returns>
+        internal static CompletionItemRules Create(
+            ImmutableArray<CharacterSetModificationRule> filterCharacterRules,
+            ImmutableArray<CharacterSetModificationRule> commitCharacterRules,
+            EnterKeyRule enterKeyRule,
+            bool formatOnCommit,
+            bool preselect)
+        {
+            var matchPriority = preselect ? Completion.MatchPriority.Preselect : Completion.MatchPriority.Default;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: can be simplified to: return Create(filter, commit, enter, format, preselect ? ... : ...)

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionItemRules.cs]: then we're doing nothing but forwarding to the new method.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
retest vsi please

**Comment by rchande** [comment]
@CyrusNajmabadi @paulvanbrenk What do we need to do to confirm that TS will be able to consume this with a recompile?

**Comment by rchande** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
retest  windows_vsi_p3_open_prtest  please
